netbox/netbox/api/pagination.py

```diff
@@ -83,3 +87,167 @@ def get_queryset_count(self, queryset):
+                kwargs = {order_attr + '__lt': current_position}
+            else:
+                kwargs = {order_attr + '__gt': current_position}
+
+            queryset = queryset.filter(**kwargs)
+
+        if self.page_size:
+            # If we have an offset cursor then offset the entire page by that amount.
+            # We also always fetch an extra item in order to determine if there is a
+            # page following on from this one.
+            results = list(queryset[offset:offset + self.page_size + 1])
+            self.page = list(results[:self.page_size])
+        else:
+            self.page = results = list(queryset[offset:])
+
+        # Determine the position of the final item following the page.
+        if len(results) > len(self.page):
+            has_following_position = True
+            following_position = self._get_position_from_instance(results[-1], self.ordering)
+        else:
+            has_following_position = False
+            following_position = None
+
+        if reverse:
+            # If we have a reverse queryset, then the query ordering was in reverse
+            # so we need to reverse the items again before returning them to the user.
+            self.page = list(reversed(self.page))
+
+            # Determine next and previous positions for reverse cursors.
+            self.has_next = (current_position is not None) or (offset > 0)
+            self.has_previous = has_following_position
+            if self.has_next:
+                self.next_position = current_position
+            if self.has_previous:
+                self.previous_position = following_position
+        else:
+            # Determine next and previous positions for forward cursors.
+            self.has_next = has_following_position
+            self.has_previous = (current_position is not None) or (offset > 0)
+            if self.has_next:
+                self.next_position = following_position
+            if self.has_previous:
+                self.previous_position = current_position
+
+        # Display page controls in the browsable API if there is more
+        # than one page.
+        if (self.has_previous or self.has_next) and self.template is not None:
+            self.display_page_controls = True
+
+        return self.page
+
+    def get_page_size(self, request):
+        if self.page_size_query_param:
+            try:
+                page_size = int(request.query_params[self.page_size_query_param])
+                if page_size < 0:
+                    raise ValueError()
+                # Enforce maximum page size, if defined
+                MAX_PAGE_SIZE = get_config().MAX_PAGE_SIZE
+                if MAX_PAGE_SIZE:
+                    return MAX_PAGE_SIZE if page_size == 0 else min(page_size, MAX_PAGE_SIZE)
+                return page_size
+            except (KeyError, ValueError):
+                pass
+
+        return self.default_page_size
+
+
+class TwoModePagination(BasePagination):
```

Comment:
I understand why this class was introduced, given that DRF does not support dynamic paginator selection, but I'm afraid this approach isn't tenable. We need to keep the paginator construction relatively simple to ensure its long-term maintainability, and using a sort of proxy class to choose between two different paginators is far from ideal.

Instead, we should be able to override the `paginator()` method of GenericAPIView to select one class or the other based on the value of the `pagination_mode` parameter. (`self.request` is accessible when `paginator()` is called.) I

If that ends up not being feasible, perhaps a reasonable compromise would be expose the paginator as a NetBox configuration parameter. This would allow an administrator to designate either pagination strategy application-wide, however it would not be user-selectable.